### ISSUE TITLE
DefaultMockServer alters binary data/byte arrays/octet stream

### ISSUE THREAD

**Issue by Steve973** [fabric8io/kubernetes-client#6061]
### Describe the bug

I am creating a basic, non-TLS `DefaultMockServer` for some testing that I am doing.  I set up a byte array for an interaction/expect to return as a response and retrieve it with an `HttpURLConnection`.  I examine the data that I am sending back from the web server, and I examine the data that I actually received, and they *do not match*!  I was looking for documentation that describes this behavior and, either it does not exist, or this is not the intended behavior.  I think that this should absolutely not be the intended behavior.

### Fabric8 Kubernetes Client version

6.13.0

### Steps to reproduce

  I have created code that demonstrates this (mis)behavior simply, consistently and repeatably:

```java
    void testBytesRoundTrip() throws Exception {
        BiConsumer<String, byte[]> printBytes = (desc, bytes) -> {
            StringBuilder sb = new StringBuilder();
            for (Byte b : bytes) {
                sb.append(String.format("%02x", b));
            }
            System.err.println("########## " + desc + " (bytes as hex): " + sb);
        };
        DefaultMockServer mockServer = new DefaultMockServer();
        mockServer.start();
        String mockServerUrl = mockServer.url("");
        byte[] dataBytes = new byte[]{0x01, 0x02, 0x03, 0x04};
        printBytes.accept("Sending data", dataBytes);
        HashMap<String, String> headers = new HashMap<>();
        headers.put("Content-Type", "application/octet-stream");
        mockServer.expect()
                .post()
                .withPath("/")
                .andReply(ResponseProviders.of(200, dataBytes, headers))
                .once();

        HttpURLConnection con;
        URL url = new URL(mockServerUrl);
        con = (HttpURLConnection) url.openConnection();
        con.setDoOutput(true);

        try (OutputStream out = con.getOutputStream();
             DataOutputStream dataOut = new DataOutputStream(new BufferedOutputStream(out))) {
            dataOut.write(dataBytes);
            dataOut.flush();
        }

        int responseCode = con.getResponseCode();
        if (responseCode >= 200 && responseCode < 300) {
            Map<String, List<String>> rxHeaders = con.getHeaderFields();
            for (Map.Entry<String, List<String>> entry : rxHeaders.entrySet()) {
                System.err.println("Received header: " + entry.getKey() + " -> " + entry.getValue());
            }
            int contentLength = con.getContentLength();
            try (InputStream in = con.getInputStream();
                 ByteArrayOutputStream out = new ByteArrayOutputStream(contentLength)) {
                int readByte;
                while ((readByte = in.read()) != -1) {
                    out.write(readByte);
                }
                out.flush();
                printBytes.accept("Received data", out.toByteArray());
            }
        }
        mockServer.shutdown();
    }
```
here is the output:
```text
########## Sending data (bytes as hex): 01020304
Jun 18, 2024 11:28:10 AM okhttp3.mockwebserver.MockWebServer$2 execute
INFO: MockWebServer[57777] starting to accept connections
Jun 18, 2024 11:28:10 AM okhttp3.mockwebserver.MockWebServer$3 processOneRequest
INFO: MockWebServer[57777] received request: POST / HTTP/1.1 and responded: HTTP/1.1 200 OK
Received header: null -> [HTTP/1.1 200 OK]
Received header: Content-Length -> [10]
Received header: Content-Type -> [application/octet-stream]
########## Received data (bytes as hex): 224151494442413d3d22
Jun 18, 2024 11:28:10 AM okhttp3.mockwebserver.MockWebServer$2 acceptConnections
INFO: MockWebServer[57777] done accepting connections: Socket closed
```

### Expected behavior

The mock server should return the exact data that was set up in the expectation, regardless of how the underlying implementation handles the data internally.

### Runtime

other (please specify in additional context)

### Kubernetes API Server version

other (please specify in additional context)

### Environment

Linux

### Additional context

This is just the fabric8 DefaultMockServer without Kubernetes, although that is probably entirely apparent from the problem description, but it asked me to specify details in this field.

**Comment by manusa** [comment]
I don't think that the current MockServer implementations supports responses that can't be converted to Strings.

In version 7.0.0 we're going to switch the core library supporting the Kubernetes Mock Server implementation from OkHttp to Vert.x.

There is a PoC for the new implementation at https://github.com/fabric8io/kubernetes-client/pull/5632.
You might want to check it out and see if it covers your use case. If you do, please provide feedback.

**Comment by Steve973** [comment]
Ok.  I just made a node in the comments to please make sure that the data arrives unmolested (though in less pedestrian wording).

**Comment by manusa** [comment]
> I cannot even find where (in either fabric8 code, or in okhttp3 code

This is probably part of the OkHttp implementation when the body of the mock response is set as a string (either in the mockwebserver or okio modules).

>  I just made a node in the comments 

Have you actually tried the reproducer with the compiled changes in the PoC?

**Comment by Steve973** [comment]
> > I cannot even find where (in either fabric8 code, or in okhttp3 code
> 
> This is probably part of the OkHttp implementation when the body of the mock response is set as a string (either in the mockwebserver or okio modules).
I'm not setting anything as a String, so if that's the likely cause, then something else must be going on.

> Have you actually tried the reproducer with the compiled changes in the PoC?
Not yet.  Do these builds get deployed to any repo? Or is this something that I will need to build for myself?

**Comment by manusa** [comment]
> I'm not setting anything as a String, so if that's the likely cause, then something else must be going on.

I know, but the original Kubernetes Mock Server implementation most likely is.

> Not yet. Do these builds get deployed to any repo? Or is this something that I will need to build for myself?

Not for the pull requests. For this one you'd need to check out the branch, and install to your local Maven repo (`mvn -DskipTests clean install`). Then, update the dependency in your project to `7.0-SNAPSHOT`)

**Comment by Steve973** [comment]
I can use a Buffer with the okhttp3 version (that fabric8 is using) and it doesn't change anything about the body, so this is something that fabric8 is introducing at least by the usage, if not through specific code.  This comment is just a bit of an aside, or FYI, and not related to building and using the snapshot.

**Comment by manusa** [comment]
Hi @Steve973,
Could you please check if this is still going on with the latest v7.0.1 release of the Mock Server?
